gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/endpoint/FidoFactorUtils.java

```diff
@@ -0,0 +1,124 @@
+/**
+ * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.gravitee.am.gateway.handler.root.resources.endpoint;
+
+import io.gravitee.am.common.factor.FactorType;
+import io.gravitee.am.gateway.handler.common.factor.FactorManager;
+import io.gravitee.am.gateway.handler.root.service.user.UserService;
+import io.gravitee.am.identityprovider.api.DefaultUser;
+import io.gravitee.am.model.Factor;
+import io.gravitee.am.model.User;
+import io.gravitee.am.model.factor.EnrolledFactor;
+import io.gravitee.am.model.factor.EnrolledFactorSecurity;
+import io.gravitee.am.model.factor.FactorStatus;
+import io.gravitee.am.model.oidc.Client;
+import io.reactivex.Completable;
+import io.vertx.reactivex.ext.web.RoutingContext;
+import io.vertx.reactivex.ext.web.Session;
+
+import java.util.Date;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+import static io.gravitee.am.common.factor.FactorSecurityType.WEBAUTHN_CREDENTIAL;
+import static io.gravitee.am.common.factor.FactorType.FIDO2;
+import static io.gravitee.am.common.utils.ConstantKeys.ENROLLED_FACTOR_ID_KEY;
+
+/**
+ * @author Ashraful Hasan (ashraful.hasan at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class FidoFactorUtils {
+    private final FactorManager factorManager;
+    private final UserService userService;
+
+    public FidoFactorUtils(FactorManager factorManager, UserService userService) {
+        this.factorManager = factorManager;
+        this.userService = userService;
+    }
+
+    private EnrolledFactor createEnrolledFactor(String factorId, String credentialId) {
+        final EnrolledFactor enrolledFactor = new EnrolledFactor();
+        enrolledFactor.setFactorId(factorId);
+        enrolledFactor.setStatus(FactorStatus.ACTIVATED);
+        enrolledFactor.setCreatedAt(new Date());
+        enrolledFactor.setUpdatedAt(enrolledFactor.getCreatedAt());
+        enrolledFactor.setSecurity(new EnrolledFactorSecurity(WEBAUTHN_CREDENTIAL, credentialId));
+
+        return enrolledFactor;
+    }
+
+    public Completable saveFidoFactor(User user, RoutingContext routingContext, String credentialId) {
+        final Session session = routingContext.session();
+        final String fidoFactorId = session.get(ENROLLED_FACTOR_ID_KEY);
```

Comment:
`fidoFactorId` maybe null here, so the `factorManager.geFactor` will return null and NullPointer will be rised at line 70